### PR TITLE
Add mlplatform jupyterhub, mlflow, deploy resources

### PR DESCRIPTION
**Pull request by schirevko**

### CODE REVIEW

File: vkcs/mlplatform/resource_deploy.go [hunk 1]
```diff
@@ -0,0 +1,542 @@
+package mlplatform
```
Comment (ftersin) [inline vkcs/mlplatform/resource_deploy.go:1]: lets rename the file to fit naming schema

File: vkcs/mlplatform/resource_jupyterhub.go [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+				Description: "The availability zone in which to create the resource. Changing this creates a new resource",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+
+			"boot_volume": schema.SingleNestedAttribute{
+				Attributes: map[string]schema.Attribute{
+					"size": schema.Int64Attribute{
+						Required:    true,
```
Comment (ftersin) [inline vkcs/mlplatform/resource_jupyterhub.go:120]: probably optional (to let us get default one from corresponding image)?

File: vkcs/mlplatform/resource_jupyterhub.go [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+					},
+				},
+				Required:    true,
+				Description: "Instance's root volume configuration",
+			},
+
+			"data_volumes": schema.ListNestedAttribute{
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"size": schema.Int64Attribute{
```
Comment (ftersin) [inline vkcs/mlplatform/resource_jupyterhub.go:149]: could we specify this set once to reuse it?

File: vkcs/mlplatform/resource_jupyterhub.go [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+				Optional:    true,
+				Description: "Keypair name",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+
+			"s3fs_bucket": schema.StringAttribute{
+				Optional:    true,
+				Description: "S3 Bucket",
```
Comment (ftersin) [inline vkcs/mlplatform/resource_jupyterhub.go:208]: s3 bucket for what?

File: vkcs/mlplatform/resource_jupyterhub.go [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+					},
+				},
+				Required:    true,
+				Description: "Network configuration",
+				Validators: []validator.List{
+					listvalidator.SizeAtMost(1),
+				},
+			},
+
+			"keypair": schema.StringAttribute{
```
Comment (ftersin) [inline vkcs/mlplatform/resource_jupyterhub.go:198]: do not forget to remove this

File: vkcs/mlplatform/resource_mlflow.go [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+	FlavorID         types.String              `tfsdk:"flavor_id"`
+	JHInstanceID     types.String              `tfsdk:"jh_instance_id"`
+	AvailabilityZone types.String              `tfsdk:"availability_zone"`
+	BootVolume       *MLPlatformVolumeModel    `tfsdk:"boot_volume"`
+	DataVolumes      []*MLPlatformVolumeModel  `tfsdk:"data_volumes"`
+	Networks         []*MLPlatformNetworkModel `tfsdk:"networks"`
+	DemoMode         types.Bool                `tfsdk:"demo_mode"`
+	Keypair          types.String              `tfsdk:"keypair"`
+	CreatedAt        types.String              `tfsdk:"created_at"`
+	PublicIP         types.String              `tfsdk:"public_ip"`
```
Comment (ftersin) [inline vkcs/mlplatform/resource_mlflow.go:52]: does it make sence for mlflow types?

File: vkcs/mlplatform/resource_jupyterhub.go [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+				Optional:    true,
+				Description: "Keypair name",
```
Comment (paaanic) [inline vkcs/mlplatform/resource_jupyterhub.go:200]: Changing this creates a new resource

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 4 above)
Comment (paaanic) [inline vkcs/mlplatform/resource_jupyterhub.go:208]: Changing this creates a new resource

File: vkcs/mlplatform/resource_deploy.go [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+
+			"timeouts": timeouts.Attributes(ctx, timeouts.Opts{
+				Create: true,
+				Delete: true,
+			}),
+		},
+		Description: "Manages a mlplatform Deploy resource.",
```
Comment (paaanic) [inline vkcs/mlplatform/resource_deploy.go:240]: I think it's better to use "ML Platform" everywhere since it's service name defined in cloud.

**Review by paaanic — CHANGES_REQUESTED** [review]

File: vkcs/mlplatform/resource_deploy.go (hunk 1 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: renamed

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 3 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_jupyterhub.go]: moved common schema attributes to shared.go

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 4 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_jupyterhub.go]: expanded description

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 5 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_jupyterhub.go]: removed

File: vkcs/mlplatform/resource_mlflow.go (hunk 6 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_mlflow.go]: removed publicIP

File: vkcs/mlplatform/resource_deploy.go [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+
+			"region": schema.StringAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "The `region` to fetch availability zones from, defaults to the provider's `region`.",
+				PlanModifiers: []planmodifier.String{
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: Moved planModifier from kubernetes_addon to common place and reused it here

File: vkcs/mlplatform/resource_deploy.go [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+
+	bootVolume, dataVolumes, availabilityZone := flattenVolumeOpts(deployResp.Volumes)
+	data.BootVolume = bootVolume
+	sort.Slice(dataVolumes, func(i, j int) bool {
+		return dataVolumes[i].Name.ValueString() < dataVolumes[j].Name.ValueString()
+	})
+	data.DataVolumes = dataVolumes
+	data.AvailabilityZone = types.StringValue(availabilityZone)
+
+	// Save data into Terraform state
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: Removed

File: vkcs/mlplatform/resource_deploy.go [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+				Optional:    true,
+				Computed:    true,
+				Description: "The `region` to fetch availability zones from, defaults to the provider's `region`.",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+
+			"timeouts": timeouts.Attributes(ctx, timeouts.Opts{
+				Create: true,
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: Added

File: vkcs/mlplatform/resource_deploy.go [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+		return
+	}
+
+	diags = req.State.Get(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	timeout, diags := data.Timeouts.Create(ctx, instanceCreateTimeout)
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: Fixed

File: vkcs/mlplatform/resource_deploy.go [hunk 13]
```diff
@@ -0,0 +1,542 @@
...
+func (r *DeployResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var data DeployResourceModel
+
+	diags := req.State.Get(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	timeout, diags := data.Timeouts.Create(ctx, instanceDeleteTimeout)
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: Fixed

File: vkcs/mlplatform/resource_deploy.go [hunk 14]
```diff
@@ -0,0 +1,542 @@
...
+	mlPlatformClient, err := r.config.MLPlatformV1Client(region)
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating VKCS mlplatform client", err.Error())
+		return
+	}
+
+	deployID := data.ID.ValueString()
+
+	err = instances.Delete(mlPlatformClient, deployID).ExtractErr()
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: Fixed

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 7 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_jupyterhub.go]: Added

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 4 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_jupyterhub.go]: Added

File: vkcs/mlplatform/resource_deploy.go (hunk 8 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_deploy.go]: Rewrote

File: examples/mlplatform/jupyterhub/main.tf [hunk 15]
```diff
@@ -0,0 +1,26 @@
+resource "vkcs_mlplatform_jupyterhub" "jupyterhub" {
```
Comment (paaanic) [inline examples/mlplatform/jupyterhub/main.tf:1]: Let's format it with `terraform fmt`.

Comment (schirevko) [reply examples/mlplatform/jupyterhub/main.tf:1]: Done

File: docs/resources/mlplatform_jupyterhub.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+---
+
+# vkcs_mlplatform_jupyterhub
+
+Manages a ML Platform JupyterHub resource.
+
+## Example Usage
+```terraform
+resource "vkcs_mlplatform_jupyterhub" "jupyterhub" {
+    name = "tf-example"
```
Comment (ftersin) [inline docs/resources/mlplatform_jupyterhub.md:16]: format examples

File: docs/resources/mlplatform_jupyterhub.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    admin_name = "admin"
+    admin_password = "dM8Ao21,0S264iZp"
```
Comment (ftersin) [inline docs/resources/mlplatform_jupyterhub.md:18]: does ml require strict password or we can use here just "secret"

File: templates/mlplatform/resources/vkcs_mlplatform_mlflow_deploy/import.sh [hunk 18]
```diff
@@ -0,0 +1 @@
+terraform import vkcs_mlplatform_jupyterhub.myjupyterhub 0cade671-81b5-43c5-83e1-2a659378d53a
```
Comment (ftersin) [inline templates/mlplatform/resources/vkcs_mlplatform_mlflow_deploy/import.sh:1]: jupyterhub?

File: templates/mlplatform/resources/vkcs_mlplatform_mlflow/import.sh [hunk 19]
```diff
@@ -0,0 +1 @@
+terraform import vkcs_mlplatform_jupyterhub.myjupyterhub b22e0fd6-f7e5-41b7-9cc5-15c1f352bc0d
```
Comment (ftersin) [inline templates/mlplatform/resources/vkcs_mlplatform_mlflow/import.sh:1]: jupyterhub?

File: templates/mlplatform/resources/vkcs_mlplatform_jupyterhub/import.sh [hunk 20]
```diff
@@ -0,0 +1 @@
+terraform import vkcs_mlplatform_jupyterhub.myjupyterhub 3a679dd9-0942-49b0-b233-95de5a5a9502
```
Comment (ftersin) [inline templates/mlplatform/resources/vkcs_mlplatform_jupyterhub/import.sh:1]: add new line

File: vkcs/internal/services/mlplatform/v1/instances/requests.go [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+	InstanceType           string             `json:"instance_type" required:"true"`
+	JHAdminName            string             `json:"jh_admin_name,omitempty"`
+	JHAdminPassword        string             `json:"jh_admin_password,omitempty"`
+	MLFlowJHInstanceID     string             `json:"mlflow_jh_instance_id,omitempty"`
+	ISMLFlowDemoMode       bool               `json:"is_mlflow_demo_mode,omitempty"`
+	DeployMLFlowInstanceID string             `json:"deploy_mlflow_instance_id,omitempty"`
+	Flavor                 string             `json:"flavor" required:"true"`
+	Volumes                []VolumeCreateOpts `json:"volumes" required:"true"`
+	Networks               NetworkCreateOpts  `json:"networks" required:"true"`
+	Keypair                string             `json:"keypair,omitempty"`
```
Comment (ftersin) [inline vkcs/internal/services/mlplatform/v1/instances/requests.go:26]: do we really need it here?

File: vkcs/mlplatform/resource_jupyterhub.go [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+	PrivateIP        types.String              `tfsdk:"private_ip"`
+	DNSName          types.String              `tfsdk:"dns_name"`
+	Region           types.String              `tfsdk:"region"`
+	Timeouts         timeouts.Value            `tfsdk:"timeouts"`
+}
+
+func getSchemaJupyterHub(ctx context.Context, resp *resource.SchemaResponse) map[string]schema.Attribute {
+	jupyterHubAttrs := map[string]schema.Attribute{
+		"domain_name": schema.StringAttribute{
+			Optional:    true,
```
Comment (ftersin) [inline vkcs/mlplatform/resource_jupyterhub.go:63]: do not we need here Compute = true as well?

File: vkcs/mlplatform/resource_mlflow.go [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+			if planDataVolume.VolumeID == dataDataVolume.VolumeID && planDataVolume.Size != dataDataVolume.Size {
+				changedVolumes = append(changedVolumes, instances.ResizeVolumeParams{
+					ID:   planDataVolume.VolumeID.ValueString(),
+					Size: int(planDataVolume.Size.ValueInt64()),
+				})
+			}
+		}
+	}
+
+	if len(changedVolumes) > 0 {
```
Comment (ftersin) [inline vkcs/mlplatform/resource_mlflow.go:327]: could we commonize this section?

File: docs/resources/mlplatform_jupyterhub.md (hunk 16 above)
Comment (schirevko) [reply docs/resources/mlplatform_jupyterhub.md]: Done

File: docs/resources/mlplatform_jupyterhub.md (hunk 17 above)
Comment (schirevko) [reply docs/resources/mlplatform_jupyterhub.md]: Some validation is present, change sample password to "Password!"

File: templates/mlplatform/resources/vkcs_mlplatform_mlflow_deploy/import.sh (hunk 18 above)
Comment (schirevko) [reply templates/mlplatform/resources/vkcs_mlplatform_mlflow_deploy/import.sh]: Fixed

File: templates/mlplatform/resources/vkcs_mlplatform_mlflow/import.sh (hunk 19 above)
Comment (schirevko) [reply templates/mlplatform/resources/vkcs_mlplatform_mlflow/import.sh]: Fixed

File: templates/mlplatform/resources/vkcs_mlplatform_jupyterhub/import.sh (hunk 20 above)
Comment (schirevko) [reply templates/mlplatform/resources/vkcs_mlplatform_jupyterhub/import.sh]: Done

File: vkcs/internal/services/mlplatform/v1/instances/requests.go (hunk 21 above)
Comment (schirevko) [reply vkcs/internal/services/mlplatform/v1/instances/requests.go]: Removed

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 22 above)
Comment (schirevko) [reply vkcs/mlplatform/resource_jupyterhub.go:62]: This field is used only for creation; result is stored in dns_name attribute

**Review by paaanic — APPROVED** [review]

File: vkcs/mlplatform/resource_jupyterhub.go (hunk 2 above)
Comment (ftersin) [reply vkcs/mlplatform/resource_jupyterhub.go:90]: decided to postpone

File: vkcs/mlplatform/resource_mlflow.go (hunk 23 above)
Comment (ftersin) [reply vkcs/mlplatform/resource_mlflow.go:320]: decided to postpone

**Pull request merged**